### PR TITLE
JAMES-2774 Using another docker rabbit for the unstable test

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

**Comment by Arsnael** [comment]
I'm just trying to understand in what this is supposed to work better than the docker that restarts at every test? (which didn't solve the issue)

**Comment by trantienduchn** [comment]
> I'm just trying to understand in what this is supposed to work better than the docker that restarts at every test? (which didn't solve the issue)

Simple, I don't restart docker in every tests like before we did. Except the only unstable test, I start a new docker rabbitmq to serve only that test. And it suppose to work well because now there is no test data isolation violated to that unstable test.

**Comment by trantienduchn** [comment]
no space left, test this please

**Comment by chibenwa** [comment]
> Simple, I keep docker running (no restarting) in every tests like before we did. Except the only unstable test, I start a new docker rabbitmq to serve only that test. And it suppose to work well because now there is no test data isolation violated to that unstable test.

AND we are faster ^^

**Comment by Arsnael** [comment]
Ok let's give it a shot then !

**Comment by trantienduchn** [comment]
:green_apple:

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java [hunk 1]
```diff
@@ -115,8 +116,23 @@ public void afterEach(ExtensionContext context) {
...
     private Mono<Connection> resilientConnection;
 
+    /**
+     * Each test class creates it own an extension,
+     * therefore number of classes = number of extensions = number of docker rabbitmq
+     * At some moments there are 5 docker rabbitmq are up.
+     *
+     * By not using @RegisteredExtension, and triggering extension execution
+     * at beforeAll, beforeEach, afterEach, afterAll, we can avoid starting many docker rabbitmq
+     */
```
Comment (Arsnael) [inline mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java]: How about a using a Singleton then?

**Comment by trantienduchn** [comment]
after
```
[ea3350158de82c09e01d554af08a34aad0a64b69] [INFO] Apache James :: Mailbox :: Event :: RabbitMQ implementation SUCCESS [16:23 min]
```

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java [hunk 2]
```diff
@@ -144,6 +160,13 @@ void tearDown() {
             .forEach(queueName -> sender.delete(QueueSpecification.queue(queueName.asString())).block());
         sender.delete(ExchangeSpecification.exchange(MAILBOX_EVENT_EXCHANGE_NAME)).block();
         sender.close();
+
+        rabbitMQExtension.afterEach(null);
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java]: Extract `null` in a constant named `NO_EXTENSION_CONTEXT`

**Comment by trantienduchn** [comment]
recorded from old build
```
Apache James RabbitMQ backend ...................... SUCCESS [01:24 min]
Apache James :: Server :: Mail Queue :: RabbitMQ ... SUCCESS [01:49 min]
Apache James :: Mailbox :: Event :: RabbitMQ implementation SUCCESS [16:23 min]
```

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java (hunk 2 above)
Comment (trantienduchn) [reply mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java]: I will remove this commit and do what Rene proposed, seems like it is cleaner than this way.

**Comment by trantienduchn** [comment]
recorded from new build:
```
Apache James RabbitMQ backend ...................... SUCCESS [01:08 min]
Apache James :: Server :: Mail Queue :: RabbitMQ ... SUCCESS [01:19 min]
Apache James :: Mailbox :: Event :: RabbitMQ implementation SUCCESS [16:21 min]
```

**Comment by chibenwa** [comment]
:green_apple: before CI restart

**Comment by trantienduchn** [comment]
I want to improve it a bit more

**Comment by trantienduchn** [comment]
```
[d5f88f453578083198fed0a8141b529c77763633] [ERROR] dispatchShouldWorkAfterNetworkIssuesForOldRegistration  Time elapsed: 0.013 s  <<< ERROR!
[d5f88f453578083198fed0a8141b529c77763633] java.lang.IllegalStateException: Mapped port can only be obtained after the container is started
```

**Comment by trantienduchn** [comment]
```
[db4d6975efc99b7885cf4caac926a93c66dbc89c] [ERROR] Errors: 
[db4d6975efc99b7885cf4caac926a93c66dbc89c] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
[db4d6975efc99b7885cf4caac926a93c66dbc89c] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
[db4d6975efc99b7885cf4caac926a93c66dbc89c] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
[db4d6975efc99b7885cf4caac926a93c66dbc89c] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
[db4d6975efc99b7885cf4caac926a93c66dbc89c] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
[db4d6975efc99b7885cf4caac926a93c66dbc89c] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
```

**Comment by trantienduchn** [comment]
test this please

File: backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java [hunk 3]
```diff
@@ -84,14 +91,16 @@ public RabbitMQExtension build() {
...
     private final DockerRabbitMQ rabbitMQ;
     private final Consumer<DockerRabbitMQ> afterAllStage;
+    private final boolean restartDockerEachTest;
 
     private SimpleChannelPool simpleChannelPool;
     private RabbitMQConnectionFactory connectionFactory;
     private SimpleConnectionPool connectionPool;
 
-    public RabbitMQExtension(DockerRabbitMQ rabbitMQ, Consumer<DockerRabbitMQ> afterAllStage) {
+    public RabbitMQExtension(DockerRabbitMQ rabbitMQ, Consumer<DockerRabbitMQ> afterAllStage, boolean restartDockerEachTest) {
```
Comment (blackheaven) [inline backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java]: I'm a bit reluctant of using a boolean for that, but I have no simple solution for that, maybe an enum.

File: backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java [hunk 4]
```diff
@@ -32,19 +34,108 @@
...
+            }
+
+            public RabbitMQExtension build() {
+                return new RabbitMQExtension(rabbitMQ, dockerRestartPolicy);
+            }
+        }
+    }
+
+    public static Builder.RequireDockerRabbitMQ builder() {
+        return dockerRabbitMQ -> afterAllStage -> new Builder.ReadyToBuild(dockerRabbitMQ, afterAllStage);
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java]: afterAllStage -> restartPolicy

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by trantienduchn** [comment]
master:
```
Apache James RabbitMQ backend ...................... SUCCESS [01:27 min]
Apache James :: Server :: Mail Queue :: RabbitMQ ... SUCCESS [01:55 min]
Apache James :: Mailbox :: Event :: RabbitMQ implementation SUCCESS [03:13 min]
```

this change:
```
Apache James RabbitMQ backend ...................... SUCCESS [01:11 min]
Apache James :: Server :: Mail Queue :: RabbitMQ ... SUCCESS [01:19 min]
Apache James :: Mailbox :: Event :: RabbitMQ implementation SUCCESS [01:59 min]
```

File: backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java [hunk 5]
```diff
@@ -32,19 +34,108 @@
 import org.junit.jupiter.api.extension.ParameterResolver;
 
 public class RabbitMQExtension implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback, AfterEachCallback, ParameterResolver {
-    private DockerRabbitMQ rabbitMQ;
+
+    private static final Consumer<DockerRabbitMQ> DO_NOT_THING = dockerRabbitMQ -> {};
```
Comment (remk) [inline backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java]: s/DO_NOT_THING/DO_NOTHING

File: backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java [hunk 6]
```diff
@@ -32,19 +34,108 @@
...
+
+            private final DockerRabbitMQ rabbitMQ;
+            private final DockerRestartPolicy dockerRestartPolicy;
+
+            private ReadyToBuild(DockerRabbitMQ rabbitMQ, DockerRestartPolicy dockerRestartPolicy) {
+                this.rabbitMQ = rabbitMQ;
+                this.dockerRestartPolicy = dockerRestartPolicy;
+            }
+
+            public RabbitMQExtension build() {
```
Comment (mbaechler) [inline backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java]: IMO the build step is useless in this case, we can build the extension when we choose the policy

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**